#EMAL 

EMAL is a program designed to allow for the accurate estimation of sample composition given tabular BLAST results. 

##Input
EMAL is designed to take in tabular BLAST results. To ensure that the correct format is use please compare your files to the sample below. This output can be generated by using the blast command line option of *-outfmt 6*.

*Example:*
```
4925.15330	gi|658607716|ref|NC_024385.1|	90.20	51	5	0	1	51	25294	25344	1e-09	61.9
4925.15330	gi|658607305|ref|NC_024380.1|	90.20	51	5	0	1	51	24289	24339	1e-09	61.9
4925.47748	gi|548558394|ref|NC_022518.1|	94.12	51	3	0	1	51	3364	3314	2e-14	77.8
4925.111673	gi|168229273|ref|NC_010353.1|	91.84	49	4	0	1	49	23886	23934	9e-11	65.9
```
##Usage

EMAL is a three part program that has a select number of dependencies. The different stages of EMAL are designed to be run in sequence and require that one completes prior to the initialization of the next stage. The three different parts of EMAL are:

### 1. EMAL - DataPrep
####Required Files
| File     | Description   | Possible Values|
| ---|---|---|
|all.fna|A file containing all the sequences used to BLAST the sample against| Full file path|
|gi_taxid_nucl.dmp|A file containing all the Nucleotide Identifiers and their corresponding NCBI Taxonomy Identifiers|Full file path|

####all.fna Format Example
Users can provide their own custom reference sequences to EMAL so long as they provide a file in the following format where the genome identifier is between the first and second | characters. In the example below the genome identifier is *253682848*. The file must also contain the genome sequence so that EMAL can determine the length of the sequence.

*Example:*
```
>gi|253682848|ref|NZ_ACSJ01000012.1| Clostridium phage D-1873 CLG.Contig163, whole genome shotgun sequence
CAAAAATAATACCTGTTGAATATTTATAGTCATATTGAATATATGGTGTAGTAAGCAAGTTTTTCACCTCCATTATTTTT
```
####gi_taxid_nucl.dmp Format Example
The user may again provide their own dump file if working with a custom dataset by ensuring that file is in the tabular format below. The genome identifier is presented first followed by the taxonomy identifier. In the example bellow 2, 3, and 4 are genome identifiers while 9913, 9913, and 9646 are their respective identifiers. 

*Example:*
```
2	9913
3	9913
4	9646
```

####Required Command Line Options
| Option     | Position   | Possible Values| Example |
|---|---|---|
|Path to all.fna| 1| Full file path| */path/to/all.fna*|
|Path to gi_taxid_nucl.dmp|2 |Full file path| */path/to/gi_taxid_nucl.dmp*|
|Number of threads to run|3|1 - 32| *4*|
|Output file name|4|String|*genomeDataOutput.csv*|

####Running EMAL-DataPrep

To run EMAL-DataPrep ensure that the file format guidelines above have been followed. 


 1. Open a new terminal window.
 2. Naviagte to the folder containing EMAL-DataPrep.py
 3. Run the following command:
```
python3 EMAL-DataPrep.py /path/to/all.fna /path/to/gi_taxid_nucl.dmp 4 genomeDataOutput.csv
```
EMAL-DataPrep is now running and will finish by providing a new cvs file with your choosen output name. 

###2. EMAL - Main
EMAL-Main is the main step of the program that performs the estimation of community structure. It has a number of different command line arguments.

####Required Command Line Options
| Option     | Description   | Possible Values|
|---|---|---|
|-d or --directory| The full path to the folder containing all the tabular blast results to be used in the analysis. | Full file path |
####Optional Command Line Options
| Option     | Description   | Possible Values| Default Value|
|---|---|---|
|-v or --verbose|Terminal will output statements ot indicate the stage of the process | -v | None |
|-t or --threads|How many threads EMAL can run simultaneously to process information faster. *Only parts of EMAL are multithreaded thus all threads will not always be used.* | 1-32 | 1 |
|-c or --csvname| The output name of the file where results will be placed. | *Any String* | output-ADJK.csv |
|-m or --maximumLikelihoodConvergenceCriterion| The value by which the difference between estimate steps are checked. If the difference between each value between steps is < the value then the estimation ends | Decimal Numbers| 0.0001|
|-o or --outputdir| The directory where the output file is to be placed.| Full file path | The current working directory |
|-e or --fileext|The file extension of all files contain tabular BLAST results to be processed |.tsv| .tsv|

###3. EMAL - Post